### PR TITLE
Implement set_ttl for Tcp and Udp sockets

### PR DESCRIPTION
**Pull request by dlrobertson**
- Add the `ttl` member to the `IpRepr`
 - Add the `ttl` member along with setters and getters to the tcp and udp
   socket types
 - Add unit tests for the new `set_ttl` parameter
 - Update usage of `IpRepr` to include the `ttl` value

Resolves: #58

### CODE REVIEW

File: src/socket/raw.rs [hunk 1]
```diff
@@ -270,7 +270,8 @@ mod test {
         src_addr: Ipv4Address([10, 0, 0, 1]),
         dst_addr: Ipv4Address([10, 0, 0, 2]),
         protocol: IpProtocol::Unknown(IP_PROTO),
-        payload_len: 4
+        payload_len: 4,
+        ttl: 64
```
Comment (dlrobertson) [inline src/socket/raw.rs]: This isn't ergonomic to have to specify the `ttl` member on initialization.

File: src/wire/ipv4.rs [hunk 2]
```diff
@@ -734,6 +737,17 @@ mod test {
...
+        let mut bytes = vec![0; 24];
+        bytes.copy_from_slice(&REPR_PACKET_BYTES[..]);
+        let mut packet = Packet::new(&mut bytes);
+        packet.set_version(6);
+        packet.fill_checksum();
+        let packet = Packet::new(&*packet.into_inner());
+        assert_eq!(Repr::parse(&packet, &ChecksumCapabilities::default()), Err(Error::Malformed));
+    }
+
+    #[test]
```
Comment (dlrobertson) [inline src/wire/ipv4.rs]: This is totally unrelated to the TTL value change. I'm 100% fine with backing this out.

File: src/wire/ip.rs [hunk 3]
```diff
@@ -336,6 +337,26 @@ impl IpRepr {
...
+    pub fn ttl(&self) -> u8 {
+        match self {
+            &IpRepr::Unspecified { ttl, .. } => ttl,
+            &IpRepr::Ipv4(Ipv4Repr { ttl, .. }) => ttl,
+            &IpRepr::__Nonexhaustive => unreachable!()
+        }
+    }
+
+    /// Set the TTL value.
+    pub fn set_ttl(&mut self, limit: u8) {
```
Comment (dlrobertson) [inline src/wire/ip.rs]: This isn't super useful. I'd be okay with backing this out.

File: src/socket/tcp.rs [hunk 4]
```diff
@@ -311,6 +314,23 @@ impl<'a> TcpSocket<'a> {
...
     }
 
+    /// Return the TTL value.
+    ///
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the ttl value.
```
Comment (whitequark) [inline src/socket/tcp.rs]: `Set the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.`

File: src/socket/tcp.rs [hunk 5]
```diff
@@ -311,6 +314,23 @@ impl<'a> TcpSocket<'a> {
...
+    ///
+    /// A socket without an explicitly set ttl will fallback to the default [INNA recommended]
```
Comment (whitequark) [inline src/socket/tcp.rs]: `an explicitly set TTL value`, `IANA recommended`

File: src/socket/udp.rs [hunk 6]
```diff
@@ -94,6 +97,23 @@ impl<'a, 'b> UdpSocket<'a, 'b> {
...
     }
 
+    /// Return the TTL value.
+    ///
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the ttl value.
```
Comment (whitequark) [inline src/socket/udp.rs]: Same as above.

File: src/socket/tcp.rs [hunk 7]
```diff
@@ -179,6 +179,8 @@ pub struct TcpSocket<'a> {
     timeout:         Option<u64>,
     /// Interval at which keep-alive packets will be sent.
     keep_alive:      Option<u64>,
+    /// The Hop Limit (ipv6)/TTL (ipv4) to be used for the socket
```
Comment (whitequark) [inline src/socket/tcp.rs]: `The time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.`

File: src/socket/tcp.rs [hunk 8]
```diff
@@ -311,6 +314,30 @@ impl<'a> TcpSocket<'a> {
...
+    ///
+    /// [INNA recommended]: https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml
+    /// [RFC 1122 § 3.2.1.7]: https://tools.ietf.org/html/rfc1122#section-3.2.1.7
+    pub fn set_ttl(&mut self, ttl: Option<u8>) {
+        // A host MUST NOT send a datagram with a Time-to-Live (TTL)
+        // value of 0
+        match ttl {
+            Some(0)  => {},
+            catchall => self.ttl = catchall,
+        }
```
Comment (dlrobertson) [inline src/socket/tcp.rs]: Realized it was possible to set the TTL value for sent packets to the invalid value of 0 with `set_ttl`. Added logic to ensure that the sockets configured TTL value is only changed when the given argument is `None` or `Some(x)` where `x > 0`

File: src/socket/udp.rs [hunk 9]
```diff
@@ -94,6 +97,30 @@ impl<'a, 'b> UdpSocket<'a, 'b> {
...
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    ///  * A socket without an explicitly set TTL value uses the default [INNA recommended]
+    ///    value (`64`).
+    ///  * A TTL value of 0 is invalid and will be ignored. See [RFC 1122 § 3.2.1.7].
```
Comment (dlrobertson) [inline src/socket/udp.rs]: Added documentation about the added logic and updated documentation about the default value.

File: src/socket/udp.rs [hunk 10]
```diff
@@ -418,4 +450,27 @@ mod test {
...
     }
+
+    #[test]
+    fn test_set_ttl() {
+        let mut s = socket(buffer(0), buffer(1));
+        assert_eq!(s.bind(LOCAL_END), Ok(()));
+
+        // Ensure you cannot set the TTL value to an invalid value.
+        s.set_ttl(Some(0));
+        assert_eq!(s.ttl(), None);
```
Comment (dlrobertson) [inline src/socket/udp.rs]: Added tests for the added logic.

File: src/socket/tcp.rs [hunk 11]
```diff
@@ -311,6 +314,30 @@ impl<'a> TcpSocket<'a> {
...
+    /// Return the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    ///  * A socket without an explicitly set TTL value uses the default [INNA recommended]
```
Comment (whitequark) [inline src/socket/tcp.rs]: INNA → IANA

File: src/socket/tcp.rs [hunk 12]
```diff
@@ -311,6 +314,30 @@ impl<'a> TcpSocket<'a> {
...
+    ///    value (`64`).
+    ///  * A TTL value of 0 is invalid and will be ignored. See [RFC 1122 § 3.2.1.7].
```
Comment (whitequark) [inline src/socket/tcp.rs]: TTL value of 0 is clearly user error so you should panic rather than silently ignore.

File: src/socket/udp.rs [hunk 13]
```diff
@@ -94,6 +97,30 @@ impl<'a, 'b> UdpSocket<'a, 'b> {
...
     }
 
+    /// Return the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
```
Comment (whitequark) [inline src/socket/udp.rs]: Same as above.

File: src/wire/ip.rs (hunk 3 above)
Comment (whitequark) [reply src/wire/ip.rs]: Let's leave this out, `set_payload_len` is special here.

**Comment by dlrobertson** [comment]
Good catches. Updated.

**Comment by dlrobertson** [comment]
Removed `* IPv4 time-to-live value is fixed at 64.` from the README

**Comment by whitequark** [comment]
Sorry for the delay.

**Comment by dlrobertson** [comment]
No problem
